### PR TITLE
Add bitop library

### PR DESCRIPTION
**Pull request by Lejo1**
Hope everything is compatible, so may need some build test.

### CODE REVIEW

**Comment by sfan5** [comment]
The library should certainly not be put right into the Lua source code.

**Comment by Lejo1** [comment]
@sfan5 
That's a thing I'm just not sure about, I just used the installation guide.
So, where shall I put it and how to include if not directly into the lua source code?

**Comment by SmallJoker** [comment]
Source file: Put somewhere outside the official Lua source
CMake file: If this is managed by Minetest, include the source file using relative paths. Else save your changes and patch the files when `src/CMakeLists.txt` is read/run.

wine-staging does it in a similar way: patch files + scripts that apply a chosen set of those to the official source code.

**Comment by Lejo1** [comment]
Thanks for clearing this up!
Changed it and put the file just in src

**Comment by SmallJoker** [comment]
Further thoughts: This won't work when the system-wide Lua installation is used.

File: src/script/cpp_api/s_base.cpp [hunk 1]
```diff
@@ -35,6 +35,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 
 extern "C" {
 #include "lualib.h"
+#include "bit.c"
```
Comment (sfan5) [inline src/script/cpp_api/s_base.cpp:38]: uhh

Comment (Lejo1) [reply src/script/cpp_api/s_base.cpp]: Changed it to not use it when luajit is used because of "double build"

Comment (sfan5) [reply src/script/cpp_api/s_base.cpp]: That wasn't my complaint here, `.c` files are not supposed to be included somewhere else but rather added to CMakeLists.txt like all other files.

Comment (Lejo1) [reply src/script/cpp_api/s_base.cpp]: So I need to create an extra .h file for it?

Comment (sfan5) [reply src/script/cpp_api/s_base.cpp]: You would, yes

Comment (Lejo1) [reply src/script/cpp_api/s_base.cpp]: Done

**Comment by MoNTE48** [comment]
It seems /lib/bitop would be the good place for this.
It's a bad idea to put third-party libraries directly in the minetest source code!

**Comment by Lejo1** [comment]
@MoNTE48 
Good idea, did that!

**Comment by Lejo1** [comment]
Rebased!

**Comment by Lejo1** [comment]
Closing as I realised it's not really something needed in a game.

**Comment by Ekdohibs** [comment]
I just learned about this pull request, and I'm sad to see it go... I needed it (a long time ago) in the [turtle](https://github.com/minetest-mods/turtle/blob/master/bit32.lua) mod, and bit operations might also be useful for other things as well (flag handling, etc.)

**Comment by Lejo1** [comment]
> I just learned about this pull request, and I'm sad to see it go... I needed it (a long time ago) in the turtle mod, and bit operations might also be useful for other things as well (flag handling, etc.)

Yes, but I thing this basic stuff should be done using the Lua 5.3 bit operators (#3028)

Just not sure if that is possible as we also need support for LuaJit

**Comment by red-001** [comment]
Is there any interest in merging this? It would be pretty useful API to have access to.

**Comment by sfan5** [comment]
I think it's useful.
There's a few small issues left with this PR, namely the whitespace changes should go and `cmake/Modules/FindBit.cmake` is unnecessary (it should just directly import it).

**Comment by Lejo1** [comment]
> There's a few small issues left with this PR, namely the whitespace changes should go and cmake/Modules/FindBit.cmake is unnecessary (it should just directly import it).

Like the current change?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -509,6 +509,7 @@ include_directories(
 	${LUA_INCLUDE_DIR}
 	${GMP_INCLUDE_DIR}
 	${JSON_INCLUDE_DIR}
+	${BIT_INCLUDE_DIR}
```
Comment (sfan5) [inline src/CMakeLists.txt:512]: I don't see this or the _LIBRARY var defined anywhere

Comment (Lejo1) [reply src/CMakeLists.txt:512]: See https://github.com/minetest/minetest/blob/d03adb195626069dca5765a20fdca03aaddc1669/CMakeLists.txt#L227

**Review by rubenwardy — APPROVED** [review]

**Comment by rubenwardy** [comment]
tested with:

```lua
local band, bxor = bit.band, bit.bxor
local rshift, rol = bit.rshift, bit.rol

local m = tonumber(arg and arg[1]) or 100000
if m < 2 then m = 2 end
local count = 0
local p = {}

for i=0,(m+31)/32 do p[i] = -1 end

for i=2,m do
  if band(rshift(p[rshift(i, 5)], i), 1) ~= 0 then
    count = count + 1
    for j=i+i,m,i do
      local jx = rshift(j, 5)
      p[jx] = band(p[jx], rol(-2, j))
    end
  end
end

io.write(string.format("Found %d primes up to %d\n", count, m))
```

**Comment by kilbith** [comment]
I'd prefer to go with native bitwise operators like in Lua 5.3. The code snippet above does not read well.

**Comment by appgurueu** [comment]
> I'd prefer to go with native bitwise operators like in Lua 5.3. The code snippet above does not read well.

This is hardly possible: We would either have to upgrade to Lua 5.3 (which means dropping LuaJIT and breaking backwards compatibility as 5.3 is not fully compatible with 5.1) or fork Lua 5.1 / LuaJIT to add "native" bit ops. The snippet reads well enough, especially considering that bit ops aren't going to be used very frequently.

Alternatively, you may use a preprocessor for syntactic sugar.

**TL;DR: Out of scope for this PR.**

File: src/script/cpp_api/s_base.cpp [hunk 3]
```diff
@@ -88,6 +90,11 @@ ScriptApiBase::ScriptApiBase(ScriptingType type):
 	else
 		luaL_openlibs(m_luastack);
 
+	//  Load bit lib
```
Comment (sfan5) [inline src/script/cpp_api/s_base.cpp:93]:
```suggestion
	// Load bit library
```

File: lib/bitop/bit.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
```
Comment (sfan5) [inline lib/bitop/bit.h:18]: This should carry the same license notice as bit.c

File: CMakeLists.txt [hunk 5]
```diff
@@ -261,6 +261,11 @@ endif()
 find_package(GMP REQUIRED)
 find_package(Json REQUIRED)
 find_package(Lua REQUIRED)
+if(NOT USE_LUAJIT)
+	set(BIT_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/lib/bitop)
+	set(BIT_LIBRARY bitop)
```
Comment (sfan5) [inline CMakeLists.txt:266]: One more sorry: Can you rename these to `LUA_BIT_*`?

**Review by sfan5 — APPROVED** [review]
Works otherwise

**Comment by appgurueu** [comment]
:tada:

**Comment by sorcerykid** [comment]
> I just learned about this pull request, and I'm sad to see it go... I needed it (a long time ago) in the [turtle](https://github.com/minetest-mods/turtle/blob/master/bit32.lua) mod, and bit operations might also be useful for other things as well (flag handling, etc.)

That is why I created the BitWise Operations library which is an entirely Lua-based implementation and backwards compatible with all Minetest versions (no requirement for LuaJit) going back to 0.4.x.

https://content.minetest.net/packages/sorcerykid/bitwise/
